packages/effective-module-tree/index.js

```diff
@@ -0,0 +1,155 @@
+const fs = require( 'fs' ).promises;
+const path = require( 'path' );
+const globby = require( 'globby' );
+const debug = require( 'debug' )( 'effective-module-tree' );
+const treeify = require( 'object-treeify' );
+
+/**
+ * Generator that yields all places were node will look for modules
+ * For a path like : a/node_modules/b/node_modules/c
+ * It will generate the candidates:
+ *	- a/node_modules/b/node_modules/c/node_modules
+ *	- a/node_modules/b/node_modules
+ *	- a/node_modules
+ *	- node_modules
+ *
+ * @param {string} packagePath Package path used as seed for the traversal
+ */
+const candidates = function*( packagePath ) {
+	const parts = packagePath.split( path.sep );
+	for ( let i = parts.length; i >= 0; i-- ) {
+		// This avoid generating .../node_modules/node_modules/...
+		if ( parts[ i - 1 ] === 'node_modules' ) {
+			continue;
+		}
+		yield path.join( ...parts.slice( 0, i ), 'node_modules' );
+	}
+};
+
+/**
+ * Simplify tree. Once the full tree has been discovered and linked, we need
+ * to simplify it to generate a format compatible with treeify (see https://www.npmjs.com/package/object-treeify)
+ *
+ * This transforms a tree like:
+ *
+ * {
+ * 		name: 'packageName',
+ * 		version: '1.2.3'.
+ * 		linkedDeps: [
+ * 			{ name: 'foo', version: '1.2.4', linkedDeps: [] },
+ * 			{ name: 'bar', version: '1.2.5', linkedDeps: [] }
+ * 		]
+ * }
+ *
+ * into
+ *
+ * {
+ * 		'packageName@1.2.3': {
+ * 			'foo@1.2.4': {},
+ * 			'bar@1.2.5': {}
+ * 		}
+ * }
+ *
+ * @param {object} packageDef Representation of a package
+ * @param {string[]} parents List of parents of the current package, used to avoid circular dependencies
+ */
+const simplify = ( packageDef, parents ) => {
+	// This means the dependency couldn't be found. Maybe a missing `npm install` ?
+	if ( ! packageDef ) {
+		return;
+	}
+
+	const name = `${ packageDef.name }@${ packageDef.version }`;
+
+	// Don't get caught in an infinite loop caused by circular dependencies
+	if ( parents.includes( name ) ) {
+		return { [ name ]: '[Circular]' };
+	}
+
+	if ( ! packageDef.linkedDeps ) {
+		return { [ name ]: {} };
+	}
+
+	// Keep simplifying our deps recursively
+	const parentDeps = [ ...parents, name ];
+	return {
+		[ name ]: packageDef.linkedDeps.reduce(
+			( accumulated, dep ) => ( { ...simplify( dep, parentDeps ), ...accumulated } ),
+			{}
+		),
+	};
+};
+
+/**
+ * Finds the effective dependencies tree (aka the logical tree).
+ *
+ * How this works:
+ *
+ * - Find all package.json files
+ * - Read them and store it in a map indexed by path
+ * - Iterate the map, and for each element, iterate the dependencies
+ * - For each dependency, try to find it looking at parent directories
```

Comment:
For this step, it seems like we could rely on node resolution fairly simply without needing to reimplement it.

I think this could be defined entirely with [`module.createRequire`](https://nodejs.org/docs/latest-v12.x/api/modules.html#modules_module_createrequire_filename) and `require.resolve`. Something like:

```js
const { createRequire } = require( 'module' );
const requireFromPackage = createRequire( dirname( packageJsonPath ) );
requireFromPackage.resolve( someDependency ); // Path of the dependency
```

If I've understand correctly, this would allow us to do away with some of the logic here.

Reply:
Oh that's cool, I didn't know about that!
I'll give it a go, I rather reuse what node provides than trying to reimplement it.

Reply:
After playing with this for a bit, this is what I found:

- `require.resolve()` gets the path to the _file_ that should be loaded. This is read from `main` in `package.json`, and in some packages that `main` point a file that only exists after transpilation. It sounds like adding extra friction to be able to use `effective-module-tree` (e.g you need to compile all your project and make sure packages point to the right file)

- `require.resolve.paths()` sounds promising, but it returns `null` when a package name matches a core package of node. And because we use the dependency `events` in `client/package.json`, it will make `effective-module-tree` not usable in our repo :(

I leaning towards leaving the logic as is.

Reply:
If what you need is the package and the version, you ought to be able to do:

```
const someDependencyVersion = requireFromPackage( `${ someDependency }/package.json` ).version;
```

Reply:
I need the path that actually contains the package :(

Reply:
I'd expect this to work, but this isn't blocking and I'm not clear what's simpler at this point. I _think_ we can rely on all packages having a package.json in their root:

```js
path.dirname( requireFromPackage( `${ someDependency }/package.json` ) )
```